refactor/update-image-blur-feature

abdo-essam opened the pull request.

abdo-essam committed:
Refactor: Update ContentDetectionModel and disable tflite compression

This commit refactors the `ContentDetectionModel` to simplify the probability calculation for binary classification and removes the default category score assignments when `outputSize` is 2.

Additionally, it disables tflite compression in the build configuration to prevent potential issues with model loading or performance.

abdo-essam committed:
Merge branch 'develop' of https://github.com/LondonSquad/Novix into refactor/update-image-blue-feature

abdo-essam committed:
Refactor: Streamline image moderation logic and models

This commit refactors the image moderation process for improved efficiency and clarity.

Key changes:
- Simplified `ImageModerationProcessor`:
    - `processImage` is now `shouldModerateImage`, returning a boolean.
    - Removed detailed processing results (`ProcessingResult`, `ProcessingDetails`, `FaceInfo`, `Gender`) and reasoning strings.
    - Optimized model loading and removed unnecessary mutex locking.
    - Content detection is performed first, and if positive, face detection is skipped.
    - Removed `strictMode` and associated logic for uncertain gender detection.
- Updated `ContentDetectionModel`:
    - `detectContent` now returns a simpler `ContentResult` with only `isInappropriate` boolean.
    - Removed detailed category scores and overall score from `ContentResult`.
    - Simplified probability processing and category mapping.
    - Made `DEFAULT_CONTENT_THRESHOLD` a local constant.
- Updated `ImageModerationController`:
    - Now directly uses the boolean result from `shouldModerateImage`.
    - Removed `strictMode` parameter.
    - Added a `close()` method to release processor resources.
    - Caching logic simplified.
    - Exposes `state` as `StateFlow`.
- Updated `ImageViewFilter`:
    - Adapted to the simpler `shouldModerateImage` and `ContentResult`.
    - `onModerationResult` callback now only provides a boolean.
    - Removed `strictMode` from `ImageFilterConfig`.
    - Ensures `ImageModerationController` is closed.

abdo-essam committed:
feat: Enhance probability calculation in GenderDetectionModel

This commit refactors the probability calculation within the `GenderDetectionModel` for improved clarity and precision.

Key changes:
- Removed explicit `toFloat()` conversion for `normalizedFemaleProbability` and `normalizedMaleProbability` as the division of `Double` values already results in `Double`. The final `confidence` is still cast to `Float`.
- Simplified exception handling in `GenderDetectionModel`, `FaceDetector`, and `ImageModerationProcessor` by using `_` for unused exception variables, adhering to Kotlin conventions.

abdo-essam committed:
feat: Enhance model management and error handling

This commit introduces several improvements to model management and error handling:

- Implements shared model instances for `GenderDetectionModel` and `ContentDetectionModel` to reduce memory overhead and improve initialization time.
- Introduces `initMutex` to ensure thread-safe initialization of models.
- Adds logic to switch to Firebase models when available and fall back to local assets if Firebase models fail to download or initialize.
- Clears the `ModerationCacheManager` when models are switched to ensure consistency.
- Notifies listeners of model changes to allow UI updates.
- Improves error handling in `ContentDetectionModel` and `ImageModerationController` to prevent crashes and provide more informative error messages.
- Refactors `ModelDownloadManager` to simplify model downloading and management logic.
- Addresses potential race conditions and improves resource cleanup in `ImageModerationProcessor` and `ImageViewFilter`.
- Updates `ImageViewFilter` to handle cases where moderation is disabled or fails.
- Fixes potential `NullPointerException` in `ImageModerationController` when bitmap is invalid.
- Removes redundant logging statements.

abdo-essam committed:
Merge remote-tracking branch 'origin/develop' into refactor/update-image-blue-feature

abdo-essam committed:
refactor: GenderDetectionModel.kt: Use kotlin.math.exp

london-bot[bot] committed:
Branch was auto-updated.

abdo-essam committed:
Refactor: data: Move generateHash and MovieVideoRemoteResponse

abdo-essam committed:
feat(domain): Add test case for empty query string

abdo-essam committed:
refactor(ImageModerationState): Remove unused fields

ahmeddsamii commented on feature/ImageHaramBlur/src/main/java/com/london/imageharamblur/ImageModerationProcessor.kt:
```diff
@@ -2,212 +2,197 @@ package com.london.imageharamblur
-                                maleCount++
-                                Gender.MALE
+                        genderResult.let { result ->
+                            when {
```
this is giving an error, i think it needs else branch

abdo-essam committed:
Merge branch 'develop' of https://github.com/LondonSquad/Novix into refactor/update-image-blue-feature

# Conflicts:
#	app/src/test/java/com/london/data/mapper/ActorMapperKtTest.kt
#	data/src/main/java/com/london/data/di/NetworkModule.kt

abdo-essam committed:
refactor: Update image blue feature dependencies

Refactors NetworkModule and SharedPrefsTokenProvider
to use AuthPreferences and correct service imports for
image blue feature.

abdo-essam committed:
refactor: Make ImageModeration classes internal

The ImageModerationController and ImageModerationProcessor
classes are now internal. This restricts their usage to
within the module, improving encapsulation and preventing
accidental usage from other modules.

abdo-essam committed:
feat: Make FaceDetector thread-safe and improve performance

This commit introduces the following changes to `FaceDetector`:

- **Thread-Safety:**
    - Uses `ThreadLocal` to provide each thread with its own `Interpreter` instance.
    - Employs a `Mutex` to ensure synchronized access to interpreter creation and model loading.
- **Performance:**
    - `detectFaces` is now a suspend function running on `Dispatchers.IO` for better background processing.
    - Limits interpreter threads to 1 to manage resources more effectively across multiple instances.
    - Lazily loads the model buffer.
- **Error Handling:**
    - Includes error handling for model loading.
    - Safely closes interpreters and removes them from `ThreadLocal` in the `close()` method.

Mostafa-alsaygh commented on domain/src/test/kotlin/com/london/domain/usecase/AddToRecentSearchUseCaseTest.kt:
```diff
@@ -24,14 +24,27 @@ class AddToRecentSearchUseCaseTest {
 
     @Test
     fun `should call the repository add to recent search`() = runTest {
```
add 'when' to the name

Mostafa-alsaygh commented on feature/ImageHaramBlur/src/main/java/com/london/imageharamblur/ImageModerationProcessor.kt:
```diff
@@ -2,212 +2,201 @@ package com.london.imageharamblur
+                    }
+                }
+            } catch (e: Exception) {
+                // Ignore errors
```
at least add message for the error, why?

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

BassantM committed:
Merge branch 'develop' into refactor/update-image-blue-feature

abdo-essam committed:
Remove duplicate import and update image view filter

The duplicate import of `AuthInterceptor` has been removed from `NetworkModule.kt`.

In `ImageViewFilter.kt`:
- Added a new configuration option `showCustomContentWhenBlurred` to `ImageViewFilterConfig`.
- The `ImageViewFilter` now checks this flag. If an image is blurred and `showCustomContentWhenBlurred` is true, it displays `moderatedContent()`. Otherwise, it shows the `ModeratedImage`.

Mostafa-alsaygh commented on feature/ImageHaramBlur/src/main/java/com/london/imageharamblur/ImageModerationProcessor.kt:
```diff
@@ -2,212 +2,201 @@ package com.london.imageharamblur
+                modelsInitialized = false
+                currentModelsAreFromFirebase = false
+            } else {
+                try {
```
convert to runcatching kindly all the ones that have " ignore "

abdo-essam committed:
Refactor: Use `ImageView` from design system for image display

This commit replaces direct usage of `ImageViewFilter` from the `ImageHaramBlur` feature module with the `ImageView` composable provided by the design system.

The `ImageView` in the design system now internally uses `ImageViewFilter` and introduces a new configuration option, `showTextInsteadOfBlur`. When this flag is true and an image is determined to need blurring, text indicating this will be shown as an overlay instead of blurring the image.

This change promotes consistency in image handling across the application and centralizes image filtering logic.

abdo-essam committed:
refactor(ImageView): Remove unused ImageFilterConfig

abdo-essam committed:
Merge remote-tracking branch 'origin/develop' into refactor/update-image-blur-feature

# Conflicts:
#	data/src/main/java/com/london/data/di/NetworkModule.kt

abdo-essam committed:
refactor(network): Remove unused AuthInterceptor import

abdo-essam committed:
refactor(ActorMapperKtTest): Update imports after package change

abdo-essam committed:
refactor: ImageModerationProcessor: Use runCatching

Replaces try-catch blocks with runCatching for cleaner error handling.

abdo-essam committed:
Refactor: Image moderation utils and extensions

Move bitmap utils to separate file and refactor extensions.

abdo-essam committed:
refactor: Rename BlurBitmap.kt to BitmapBlur.kt

london-bot[bot] committed:
Branch was auto-updated.

abdo-essam committed:
refactor: designSystem - presentation

Fix: Updated image blur module visibility

london-bot[bot] committed:
Branch was auto-updated.
